lib/app/service/vendor.go

```diff
@@ -400,10 +382,20 @@ func (v *vendorer) pullAndExportImages(ctx context.Context, images []string, exp
 			"failed to create %q", layersDir)
 	}
 
-	if err := exportLayers(ctx, exportDir, images, v.dockerClient,
-		log.WithField("export-directory", exportDir), parallel, progress); err != nil {
+	registryDir := layersDir
+	if dockerCacheDir != "" {
+		registryDir = dockerCacheDir
+	}
+
+	if err := exportLayers(ctx, registryDir, images, v.dockerClient,
+		log.WithField("export-directory", exportDir), parallel, forcePull, progress); err != nil {
 		return trace.Wrap(err)
 	}
+	if dockerCacheDir != "" {
+		if err := utils.CopyDirContents(dockerCacheDir, layersDir); err != nil {
```

Comment:
Yes, you are right, there is no code that excludes unnecessary docker images. And if these images were pre-populated, then they will be in the target application.
I have mentioned about this in the description of this PR.
> This registry will contain all tagged images from the previous build. But this is not a big overhead, because docker images mostly only have a change at the last level.

Reply:
I don't understand what benefit the current "include the entire cache" implementation has.  Is it significantly cleaner/faster/easier to maintain than an implementation that only vendors the requested images?  The argument "this is not a big overhead" seems like saying "its bad, but not that bad".  If we're going to take "not that bad" it needs to be in exchange for something valuable -- timeline, ease of maintenance, etc.... classic tech debt tradeoff.  That "why" hasn't been addressed yet -- that I know of.

IMO because this behavior is inconsistent (with the gravity cache and the previous  docker cache behavior) and unintuitive (other caches do not change the content of what passes through them), this design decision should be revisited.

Reply:
We can work on get actual numbers. More information on this use case:
The pce product bundles over 200 docker images. The package is +10GB and mostly is docker images.
Pushing the 10GB of docker images to the local registry takes about 20 minutes.
Between builds, almost all of the 10GB are the same as the change is usually a docker image layer or two.
We plan to populate the cache when building stable and then use the cache to start the local docker registry in PRs.
The comment about bundling some layer or two that the app will not actually use is insignificant compared to the 20 minutes we save per build.

Reply:
> We can work on get actual numbers.

Better numbers aren't needed -- 10GB / 20 minutes is the sort of info I was looking for -- it helps me understand the context of this feature while reviewing.

> The comment about bundling some layer or two that the app will not actually use is insignificant compared to the 20 minutes we save per build.

Understood.  It sounds like this feature is intended solely for PCE -- and you've collectively made the call that the present implementation is sufficient for your use case.  Given that, I would request that either:

1) The "everything in the cache is vendored" behavior is explicitly documented alongside the flag in https://github.com/gravitational/gravity/blob/master/docs/9.x/pack.md#building-a-cluster-image
2) this flag is hidden so that Gravity's others users are less likely to see it and trip over this behavior

Reply:
I added a procedure to delete unneeded images in the registry.